staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go

```diff
@@ -63,6 +63,81 @@ func TestWebhookConverterWithoutWatchCache(t *testing.T) {
 	testWebhookConverter(t, false)
 }
 
+// TestWebhookNotCalledForUnusedVersions tests scenario where conversion webhook could be called for
+// versions that are nor served or nor stored.
+// Described in detail in https://github.com/kubernetes/kubernetes/issues/129979
+func TestWebhookNotCalledForUnusedVersions(t *testing.T) {
+	// KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE allows for APIs pending removal to not block tests
+	t.Setenv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE", "true")
+
+	etcd3watcher.TestOnlySetFatalOnDecodeError(false)
+	defer etcd3watcher.TestOnlySetFatalOnDecodeError(true)
+
+	tearDown, config, options, err := fixtures.StartDefaultServer(t, "--watch-cache=true")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	apiExtensionsClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		tearDown()
+		t.Fatal(err)
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		tearDown()
+		t.Fatal(err)
+	}
+	defer tearDown()
+
+	crd := multiVersionFixture.DeepCopy()
+
+	RESTOptionsGetter := serveroptions.NewCRDRESTOptionsGetter(*options.RecommendedOptions.Etcd, nil, nil)
+	restOptions, err := RESTOptionsGetter.GetRESTOptions(schema.GroupResource{Group: crd.Spec.Group, Resource: crd.Spec.Names.Plural}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	etcdClient, _, err := storage.GetEtcdClients(restOptions.StorageConfig.Transport)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// nolint:errcheck
+	defer etcdClient.Close()
+
```

Comment:
we need to make sure that existing objects in etcd can still be successfully read and converted *from* a non-served, non-storage version, and can be updated

1. create the CRD serving all versions, storing in v1alpha2
2. server-side apply an object via the v1alpha2 API (will persist in v1alpha2)
3. update the CRD to switch the storage version to a different version, stop serving v1alpha2, and set a conversion webhook
4. ensure the conversion webhook is never asked to convert *to* v1alpha2 (as this test already does)
5. ensure the previously created object can still be *read* via a different API version
6. ensure the previously created object can still be *updated* via a different API version using server-side apply
7. ensure the update does not trigger a conversion request *to* v1alpha2

Reply:
Updated test. Server side apply with different version is failing -- looking into it.

Reply:
@sbueringer Thanks for pointing to this! Indeed, that seems to be a problem. Added cleaning of `managedFields` in test to work around this.

Reply:
> In this case, the API version does exist, even though it's no longer served, so conversion should still succeed, right?

Yes, that is correct.

> 6. ensure the previously created object can still be updated via a different API version using server-side apply
> 7. ensure the update does not trigger a conversion request to v1alpha2

I was trying to bring up the following. As long as there is v1alpha2 in managedFields an update request to the object will trigger a conversion request (I don't remember which direction). The nuance that I (and potentially @GrigoriyMikhalkin as well) missed is that you wrote "a conversion request **to** v1alpha2"

Reply:
> 7. ensure the update does not trigger a conversion request _to_ v1alpha2

ah, the presence of managedFields with v1alpha2 *may* trigger a conversion into that version solely to determine whether the incoming request conflicts with any fields owned by that version ... I can't actually remember

What we _really_ care about verifying in the test is that no list / watch requests are made to storage for the unserved version (which is what the watch cache would do). Guarding against conversions was just one way to do that

Reply:
@liggitt Yeah, basically it's testing the observed effect(call to the conversion webhook). Cleaning of managed fields is done to avoid another case when conversion is called.